c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -3676,6 +3681,141 @@ bool response_started(JSObject *self) {
 }
 } // namespace FetchEvent
 
+using TimerArgumentsVector = std::vector<JS::Heap<JS::Value>>;
+
+namespace {
+class Timer {
+public:
+  uint32_t id;
+  JS::Heap<JSObject *> callback;
+  TimerArgumentsVector arguments;
+  uint32_t delay;
+  system_clock::time_point deadline;
+  bool repeat;
+
+  Timer(uint32_t id, HandleObject callback, uint32_t delay, JS::HandleValueVector args, bool repeat)
+      : id(id), callback(callback), delay(delay), repeat(repeat) {
+    deadline = system_clock::now() + system_clock::duration(delay * 1000);
+    arguments.reserve(args.length());
+    for (auto &arg : args) {
+      arguments.push_back(JS::Heap(arg));
+    }
+  }
+
+  void trace(JSTracer *trc) {
+    JS::TraceEdge(trc, &callback, "Timer callback");
+    for (auto &arg : arguments) {
+      JS::TraceEdge(trc, &arg, "Timer callback arguments");
+    }
+  }
+};
+
+class ScheduledTimers {
+public:
+  Timer *first() {
+    if (std::empty(timers)) {
+      return nullptr;
+    } else {
+      return timers.front()
+    }
+  }
+
+private:
+  std::list<Timer *> timers;
+  static uint32_t next_id;
+
+  void add_timer(Timer *timer) {
+    auto iter = timers.begin();
+
+    for (; iter != timers.end(); iter++) {
+      if ((*iter)->deadline > timer->deadline) {
+        break;
+      }
+    }
+
+    timers.insert(iter, timer);
+  }
+
+  void repeat_first() {
+    Timer *timer = first();
+    MOZ_ASSERT(timer);
+    timer->deadline = system_clock::now() + system_clock::duration(timer->delay * 1000);
```

Comment:
Does this work? If so, it might be nice to skip having to multiply by 1000 :)
```suggestion
    timer->deadline = system_clock::now() + std::milliseconds(timer->delay);
```